### PR TITLE
feat: add update option to prevent unnecessary file updates

### PR DESCRIPTION
**Pull request by craigrmccown**
This change was inspired by the `--update` option from [cpx](https://github.com/mysticatea/cpx). Without this option, `tsm` was overwriting type definition files in my project, causing TypeScript recompilations. In a large project, this can cause some significant performance problems.

I purposefully didn't update the README or add test coverage yet because I'd like feedback on the general approach first. I'm happy to add these changes after collecting initial thoughts! Does this idea have the potential to make it into a release? Thanks!

### CODE REVIEW

File: lib/core/list-different.ts [hunk 1]
```diff
@@ -48,7 +48,7 @@ export const checkFile = (
 
       const path = getTypeDefinitionPath(file);
 
-      const content = fs.readFileSync(path, { encoding: "UTF8" });
+      const content = fs.readFileSync(path, { encoding: "utf8" });
```
Comment (craigrmccown) [inline lib/core/list-different.ts:51]: I can revert this if I need to. TypeScript wouldn't let me build the project locally without this change.

**Comment by skovy** [comment]
Hey, thanks for putting this together. Based on your explanation I think this general idea makes sense. I looked at the code but didn't fully follow from the example, would you mind walking through a specific example of saving and what happens vs what you'd like to see?

Also wondering if there's any downside to always doing this and avoid adding another option? What's a case where you wouldn't want this to happen?

**Comment by craigrmccown** [comment]
Good questions and no problem! Here's a more concrete scenario describing the current behavior:
- During local development, I want to run a `tsc` watcher process and a `tsm` watcher process, so I write a script to run both processes concurrently.
- Everything works fine - changes to SCSS modules re-generate type files, and my TypeScript is happily compiling. But, I notice it's a little bit slow to start this script.
- After some digging, I find out that when I start the script, `tsm` overwrites every generated `.d.ts` file, which causes TypeScript to recompile, causing the slowness.

It's worth noting that even if I run `tsm` without watch mode, then use `--ignoreInitial`, I run into the same problem because the files still get overwritten. Instead, if we use the mtimes of the source files compared to the output files to determine whether type declarations need to be generated, we can avoid this problem (similar to how `make` targets work).

> Also wondering if there's any downside to always doing this and avoid adding another option? What's a case where you wouldn't want this to happen?

I can't think of a scenario where we wouldn't want this to happen. In fact, for my purposes, I would always pass the `--update` option. But, I added the option so that the current behavior would remain by default, just in case other users were relying on it. If you don't think this is a concern, I would prefer to remove the option as well 😄

**Comment by skovy** [comment]
Got it, so the only use case this could cause problems is if someone manually changes a type definition file after changing the corresponding SCSS file and then runs this CLI they'll end up with invalid type definitions even after running it. Given this, I do think we should overwrite by default so the tool works out-of-the-box (although, may have some side effects like triggering a TypeScript compiler rebuild). Then, the optimizations can be opt-in with the known risks (no forceful overwrites).

I think a new option like you have makes the most sense for now. What are your thoughts on `--update`? To me, it feels a bit generic, and reading only the flag doesn't lead me to think that it's only updating "stale" files. While I prefer succinct names, I do wonder if something like `--update-stale`, `--update-only-stale`, or `--skip-fresh` would be more accurate? I'm also fine with `--update` since it's short and we can provide an in-depth explanation in the README.

So I think we can get this merged after:

- Let me know your thoughts on the flag name. I'm good with whatever you choose as long as you provide the rationale for context to myself and any others that may look at this PR in the future.
- Update the code to reflect the flag naming if it changed.
- Add tests as you mentioned.
- Add a README description as you mentioned, with a particular note of the con (may not overwrite a manually changed file) and the pro (will avoid unnecessary writes, which may be desirable if running `tsc --watch`)

File: __tests__/core/list-different.test.ts [hunk 2]
```diff
@@ -3,7 +3,7 @@ import { listDifferent } from "../../lib/core";
 import { describeAllImplementations } from "../helpers";
 
 describeAllImplementations((implementation) => {
-  describe("writeFile", () => {
+  describe("listDifferent", () => {
```
Comment (craigrmccown) [inline __tests__/core/list-different.test.ts:6]: I made an unrelated fix to a copy/paste problem that I noticed.

**Comment by craigrmccown** [comment]
Well said! RE: flag name, I agree that `--update` was too generic. I changed it to `--updateStaleOnly`. I also added tests and updated the README. Are there any further changes required before merging?

File: README.md [hunk 3]
```diff
@@ -229,6 +229,16 @@ export type IStyles = {
...
 
+### `--updateStaleOnly` (`-u`)
+
+- **Type**: `boolean`
+- **Default**: `false`
+- **Example**: `tsm src --updateStaleOnly`
+
+Overwrite generated files only if the source file has more recent changes. This can be useful if you want to avoid extraneous file updates, which can cause watcher processes to trigger unnecessarily (e.g. `tsc --watch`).
+
+Caveat: If a generated type definition file is updated manually, it won't be re-generated until the corresponding scss file is also updated.
```
Comment (skovy) [inline README.md:240]: perfect 👌

File: lib/core/list-different.ts (hunk 1 above)
Comment (skovy) [reply lib/core/list-different.ts:51]: Were you getting a type error, or run time error? The types that I have (at least locally) define this as a `string` so any string should be valid?

**Review by skovy — APPROVED** [review]
Looks great, thanks for using camel case so it's consistent 👍

**Comment by skovy** [comment]
@all-contributors please add @craigrmccown for idea, code, tests, and docs

**Comment by craigrmccown** [comment]
Thanks @skovy! I appreciate your help and quick turnaround on this.

The `UTF8` -> `utf8` change was prompted by a type error, although this could absolutely have to do with my local environment. For me, that argument was typed as a union of string literals, and only the lower-case versions were included 🤷
